code/Hi_MAP/onmt/translate/beam.py

```diff
@@ -134,15 +132,27 @@ def advance(self, word_probs, attn_out):
         # word and beam each score came from
         prev_k = best_scores_id / num_words
         self.prev_ks.append(prev_k)
+        self.prev_ks_cpu.append(prev_k.tolist())
+
         self.next_ys.append((best_scores_id - prev_k * num_words))
+        self.next_ys_cpu.append((best_scores_id - prev_k * num_words).tolist())
+
         self.attn.append(attn_out.index_select(0, prev_k))
         self.global_scorer.update_global_state(self)
-
-        for i in range(self.next_ys[-1].size(0)):
-            if self.next_ys[-1][i] == self._eos:
-                global_scores = self.global_scorer.score(self, self.scores)
-                s = global_scores[i]
+        
+        eos_indicator = self.next_ys[-1] == self._eos
+        if eos_indicator.any():
+            global_scores = self.global_scorer.score(self, self.scores)
+            global_scores_eos = global_scores[eos_indicator]
+            i_indexes = torch.where(eos_indicator)[0]
```

Comment:
I had a problem here because of my torch version, so I suggest you change this to:
`i_indexes = eos_indicator.nonzero()[0]`

Reply:
Yes, you are right. Lower torch (<1.3.0) versions don't have good bool operation support. 

How about this, `i_indexes = torch.where(eos_indicator.long())[0]` ? Could you verify it in your version?

About your non_zero suggestion, I tried
```
In [10]: print(eos_indicator)
Out[10]: tensor([False, False,  True,  True])
In [11]: torch.where(eos_indicator)[0]
Out[11]: tensor([2, 3])
In [12]: eos_indicator.nonzero()[0]
Out[12]: tensor([2])
```
The result is not same as torch.where output.

Reply:
You're correct, my bad. I'm using torch 0.4.1 (for other reasons) and by looking at the documentation, I think it's not possible to achieve the same results. There is no torch.where() with single argument in this version, and nonzero() returns the results in such a way:
`>>> torch.nonzero(eos_indicator)`
`tensor([[2],
        [3]])`
So I guess I'll just have to update my torch.

Reply:
Good to know that, then I will mark this as resolved. The current PR is compatible with majority of torch versions (>1.0).